src/main/java/hudson/matrix/MatrixProject.java

```diff
@@ -207,6 +214,24 @@ public MatrixProject(ItemGroup parent, String name) {
         super(parent, name);
     }
 
+    /**
+     * Lock the project
+     */
+    public void buildLock(){
+        // create lock instance if not exist
+        if (buildLock == null) {
+            buildLock = new ReentrantReadWriteLock();
```

Comment:
You will run into problems when 2 threads call this at the same time - they will use different lock instance and operate in parallel anyway. The first one that completed frees the lock used by latter thread for third one to enter the critical section while 2nd can still be in progress.